### PR TITLE
`static_buf!()`: Remove `StaticUninitializedBuffer` and `UninitializedBuffer` in favor of `MaybeUninit`, update component conventions to match

### PR DESCRIPTION
**Pull request by hudson-ayers**
### Pull Request Overview

This pull request builds on top of #3219 , with some additional optimizations for size to reduce the overhead introduced in #3219. This PR removes `StaticUninitializedBuffer` and `UninitializedBuffer`, since I believe they now offer very little over `MaybeUninit` today (other than a slightly nicer name for `initialize()` vs `write()`.

This combination of changes makes it no longer `unsafe` to call `static_init!()` or `static_buf!()`. A logical followup change is to remove `unsafe` from `static_buf!()` and make `Component::finalize()` safe. This change should allow us to remove >90% of the unsafe code in `main.rs`, a big jump forward for validating memory safety in Tock.

On Imix, this brings the overhead of #3219 down to 1104 bytes. Notably, once Brad finishes porting uses of `static_init_half!()` to `static_buf!()`, this overhead will increase.

Apologies for the weird commit ordering, where #3219 comes after my changes -- I implemented my changes first and then pulled in the #3219 changes in a squashed commit. Ultimately I think that #3219 should just be closed in favor of this PR.

This PR also defines the expected process for a component: _all_ statically declared memory is created in a macro and passed in to `finalize()`. That macro has a naming convention of `[name]_component_static!()`. No memory is statically allocated in `finalize()` (i.e. no static_init! or static_buf! in the finalize function). That macro _only_ statically declares memory.

### Testing Strategy

This pull request was tested by compiling.

### TODO or Help Wanted

~~After this PR is merged, Brad's various component PRs will need to be rebased to use `MaybeUninit` as the `StaticInput` to components.~~

### Documentation Updated

- [x]  Updated.

### Formatting

- [x] Ran `make prepush`.

### CODE REVIEW

**Review by bradjc — APPROVED** [review]
I think this is a major step forward for components. The singleton-without-enforcement property of components has long been a pain point for Tock, and this is a step towards eliminating that rought edge.

**Review by bradjc — DISMISSED** [review]

File: boards/opentitan/src/tests/hmac.rs [hunk 1]
```diff
@@ -72,29 +72,31 @@ impl<'a> digest::ClientVerify<32> for HmacTestCallback {
...
+    };};
 }
 
 #[test_case]
 fn hmac_check_load_binary() {
     let perf = unsafe { PERIPHERALS.unwrap() };
     let hmac = &perf.hmac;
 
-    let callback = unsafe { static_init_test_cb() };
+    let callback = unsafe { static_init_test_cb!() };
```
Comment (hudson-ayers) [inline boards/opentitan/src/tests/hmac.rs:99]: I am surprised you don't get warnings that this unsafe is not needed

**Review by bradjc — DISMISSED** [review]

File: chips/sifive/src/prci.rs [hunk 2]
```diff
@@ -59,10 +60,17 @@ impl Prci {
...
         match frequency {
-            ClockFrequency::Freq16Mhz => {}
+            ClockFrequency::Freq16Mhz => {
+                regs.pllcfg
+                    .modify(pllcfg::bypass::SET + pllcfg::refsel::SET);
+                regs.plloutdiv
+                    .modify(plloutdiv::divby1.val(1) + plloutdiv::div.val(0));
+                regs.pllcfg.modify(pllcfg::sel::SET);
+                regs.hfrosccfg.modify(hfrosccfg::enable::CLEAR);
+            }
```
Comment (alistair23) [inline chips/sifive/src/prci.rs:73]: Why is this change in here?

Comment (bradjc) [reply chips/sifive/src/prci.rs]: Rebased and fixed.

**Review by bradjc — DISMISSED** [review]

**Review by alevy — APPROVED** [review]

**Comment by alevy** [comment]
The most recent version of this PR (which @bradjc just updated) requires `unsafe` to use `static_`.

**Comment by alevy** [comment]
There was some discussion outside GitHub about the desire for a zero-cost variant that doesn't dynamically check if a value has already been initialized, though the uses of this would likely be only out-of-tree (at least for now).

I think this is pretty simple to implement, but would vote for leaving this either out-of-tree for now or doing it in a separate pull request.

**Review by bradjc — APPROVED** [review]

**Review by ppannuto — APPROVED** [review]
Got a chance to actually read this over now, sorry I wasn't so engaged earlier today. I agree this is better than what we have, and we should make the stepwise improvement here.

**Pull request merged**